File: tests_end_to_end/page_objects/TracesPage.py
from playwright.sync_api import Page, expect, Locator
import re
import time


class TracesPage:
    def __init__(self, page: Page):
        self.page = page
        self.traces_table = self.page.get_by_role("table")
        self.trace_names_selector = "tr td:nth-child(3) div span"
        self.trace_id_selector = "tr:nth-child({}) > td:nth-child(2) > div".format
        self.next_page_button_locator = (
            self.page.locator("div")
            .filter(has_text=re.compile(r"^Showing (\d+)-(\d+) of (\d+)"))
            .nth(2)
            .locator("button:nth-of-type(3)")
        )
        self.delete_button_locator = (
            self.page.locator("div")
            .filter(has_text=re.compile(r"^Add to dataset$"))
            .get_by_role("button")
            .nth(2)
        )
        self.attachments_submenu_button = self.page.get_by_role(
            "button", name="Attachments"
        )
        self.attachment_container = self.page.get_by_label("Attachments")

    def get_all_trace_names_on_page(self):
        self.page.wait_for_selector(self.trace_names_selector)
        names = self.page.locator(self.trace_names_selector).all_inner_texts()
        return names

    def click_first_trace_that_has_name(self, trace_name: str):
        self.page.get_by_role("row").filter(has_text=trace_name).first.get_by_role(
            "button"
        ).first.click()

    def check_trace_attachment(self, attachment_name=None):
        if attachment_name:
            expect(self.attachments_submenu_button).to_be_visible()
            expect(
                self.attachment_container.filter(has_text=attachment_name)
            ).to_be_visible()
        else:
            expect(self.attachments_submenu_button).to_have_count(0)

    def get_first_trace_name_on_page(self):
        self.page.wait_for_selector(self.trace_names_selector)
        name = self.page.locator(self.trace_names_selector).first.text_content()
        return name

    def get_all_trace_names_in_project(self):
        names = []
        names.extend(self.get_all_trace_names_on_page())
        while (
            self.next_page_button_locator.is_visible()
            and self.next_page_button_locator.is_enabled()
        ):
            self.next_page_button_locator.click()
            self.page.wait_for_timeout(500)
            names.extend(self.get_all_trace_names_on_page())

        return names

    def get_pagination_button(self) -> Locator:
        return self.page.get_by_role("button", name="Showing")

    def get_number_of_traces_on_page(self):
        try:
            expect(self.page.get_by_role("row").first).to_be_visible()
        except Exception:
            return 0
        finally:
            return self.page.get_by_role("row").count()

    def get_total_number_of_traces_in_project(self):
        pagination_button_text = self.get_pagination_button().inner_text()
        match = re.search(r"of (\d+)", pagination_button_text)
        if match:
            return int(match.group(1))
        else:
            return 0

    def wait_for_traces_to_be_visible(self, timeout=10, initial_delay=1):
        start_time = time.time()
        delay = initial_delay

        while time.time() - start_time < timeout:
            traces_number = self.get_number_of_traces_on_page()
            if traces_number > 0:
                return True

            time.sleep(delay)
            delay = min(delay * 2, timeout - (time.time() - start_time))
            self.page.reload()

        raise TimeoutError(f"could not get traces in UI within {timeout} seconds")

    def delete_single_trace_by_name(self, name: str):
        trace = self.page.get_by_role("row").filter(has_text=name).first
        trace.get_by_label("Select row").click()
        self.delete_button_locator.click()
        self.page.get_by_role("button", name="Delete traces").click()

    def delete_all_traces_that_match_name_contains_filter(self, name: str):
        # TODO compact this into smaller functions
        self.page.get_by_role("button", name="Filters").click()
        filter_row = self.page.get_by_role("row").filter(
            has=self.page.get_by_role("cell", name="Where")
        )
        filter_row.get_by_role("cell").filter(has_text=re.compile(r"^Column$")).click()
        self.page.get_by_label("Name").click()
        self.page.get_by_test_id("filter-string-input").click()
        self.page.get_by_test_id("filter-string-input").fill(name)
        expect(self.page.get_by_role("button", name="Filters (1)")).to_be_visible()
        self.page.keyboard.press(key="Escape")

        total_traces = self.get_total_number_of_traces_in_project()
        while total_traces > 0:
            expect(self.page.get_by_label("Select all")).to_be_visible()
            self.page.get_by_label("Select all").click()
            self.delete_button_locator.click()
            self.page.get_by_role("button", name="Delete traces").click()

            pagination_button = self.get_pagination_button()
            expect(pagination_button).not_to_have_text(
                f"Showing 1-10 of {total_traces}"
            )
            total_traces = self.get_total_number_of_traces_in_project()

    def add_all_traces_to_new_dataset(self, dataset_name: str):
        self.page.get_by_label("Select all").click()
        self.page.get_by_role("button", name="Add to dataset").click()
        self.page.get_by_role("button", name="Create new dataset").click()
        self.page.get_by_placeholder("Dataset name").fill(dataset_name)
        self.page.get_by_role("button", name="Create dataset").click()

File: tests_end_to_end/typescript-tests/page-objects/traces.page.ts
import { Page, expect, Locator } from '@playwright/test';

export class TracesPage {
  readonly page: Page;
  readonly tracesTable: Locator;
  readonly nextPageButton: Locator;
  readonly deleteButton: Locator;
  readonly attachmentsSubmenuButton: Locator;
  readonly attachmentContainer: Locator;
  private nameColumnIndex: number = -1;

  constructor(page: Page) {
    this.page = page;
    this.tracesTable = page.getByRole('table');

    this.nextPageButton = page
      .locator('div')
      .filter({ hasText: /^Showing (\d+)-(\d+) of (\d+)/ })
      .nth(2)
      .locator('button:nth-of-type(3)');

    this.deleteButton = page
      .getByRole('button')
      .filter({ has: page.locator('svg path[d*="M3 6h18"]') });

    this.attachmentsSubmenuButton = page.getByRole('button', { name: 'Attachments' });
    this.attachmentContainer = page.getByLabel('Attachments');
  }

  async initialize(): Promise<void> {
    await this.page.waitForTimeout(1000);

    // Ensure we're on the Traces view (not Threads or Spans)
    const tracesToggle = this.page.getByRole('radio', { name: 'Traces' });
    if ((await tracesToggle.getAttribute('aria-checked')) !== 'true') {
      await tracesToggle.click();
      await this.page.waitForTimeout(500);
    }

    try {
      await this.page.getByTestId('columns-button').click({ timeout: 5000 });
    } catch {
      await this.page.reload();
      await this.page.getByTestId('columns-button').click({ timeout: 5000 });
    }

    try {
      await expect(
        this.page.getByRole('button', { name: 'Name' }).getByRole('checkbox')
      ).toBeChecked({ timeout: 2000 });
    } catch {
      await this.page.getByRole('button', { name: 'Name' }).click();
    }

    await this.page.keyboard.press('Escape');
  }

  private async resolveNameColumnIndex(): Promise<number> {
    if (this.nameColumnIndex > 0) return this.nameColumnIndex;

    const headers = this.page.locator('thead th');
    const count = await headers.count();
    for (let i = 0; i < count; i++) {
      const text = await headers.nth(i).innerText();
      if (text.match(/^Name\b/)) {
        this.nameColumnIndex = i + 1;
        return this.nameColumnIndex;
      }
    }
    throw new Error('Name column not found in traces table. Was initialize() called?');
  }

  async getAllTraceNamesOnPage(): Promise<string[]> {
    const idx = await this.resolveNameColumnIndex();
    const selector = `tr td:nth-child(${idx}) div span`;
    await this.page.waitForSelector(selector);
    return await this.page.locator(selector).allInnerTexts();
  }

  async clickFirstTraceWithName(traceName: string): Promise<void> {
    await this.page.getByRole('cell', { name: traceName, exact: true }).first().click();
  }

  async checkTraceAttachment(attachmentName?: string): Promise<void> {
    if (attachmentName) {
      await expect(this.attachmentsSubmenuButton).toBeVisible();
      await expect(this.attachmentContainer.filter({ hasText: attachmentName })).toBeVisible();
    } else {
      await expect(this.attachmentsSubmenuButton).toHaveCount(0);
    }
  }

  async getFirstTraceNameOnPage(): Promise<string | null> {
    const idx = await this.resolveNameColumnIndex();
    const selector = `tr td:nth-child(${idx}) div span`;
    await this.page.waitForSelector(selector);
    return await this.page.locator(selector).first().textContent();
  }

  async getAllTraceNamesInProject(): Promise<string[]> {
    const names: string[] = [];
    names.push(...(await this.getAllTraceNamesOnPage()));

    while (
      (await this.nextPageButton.isVisible()) &&
      (await this.nextPageButton.isEnabled())
    ) {
      await this.nextPageButton.click();
      await this.page.waitForTimeout(500);
      names.push(...(await this.getAllTraceNamesOnPage()));
    }

    return names;
  }

  getPaginationButton(): Locator {
    return this.page.getByRole('button', { name: 'Showing' });
  }

  async getNumberOfTracesOnPage(): Promise<number> {
    try {
      await expect(this.page.getByRole('row').first()).toBeVisible();
      return await this.page.getByRole('row').count();
    } catch {
      return 0;
    }
  }

  async getTotalNumberOfTracesInProject(): Promise<number> {
    const paginationButtonText = await this.getPaginationButton().innerText();
    const match = paginationButtonText.match(/of (\d+)/);
    return match ? parseInt(match[1]) : 0;
  }

  async waitForTracesToBeVisible(timeout: number = 10000, initialDelay: number = 1000): Promise<void> {
    const startTime = Date.now();
    let delay = initialDelay;

    while (Date.now() - startTime < timeout) {
      const tracesNumber = await this.getNumberOfTracesOnPage();
      if (tracesNumber > 0) {
        return;
      }

      await this.page.waitForTimeout(delay);
      delay = Math.min(delay * 2, timeout - (Date.now() - startTime));
      await this.page.reload();
    }

    throw new Error(`Could not get traces in UI within ${timeout}ms`);
  }

  async deleteSingleTraceByName(name: string): Promise<void> {
    const trace = this.page.getByRole('row').filter({ hasText: name }).first();
    await trace.getByLabel('Select row').click();
    await this.deleteButton.click();
    await this.page.getByRole('button', { name: 'Delete traces' }).click();
  }
}
